From 6a6ecc746c497bddf8eb0fb38b0311bcd81d0dd2 Mon Sep 17 00:00:00 2001
From: dann frazier <dann.frazier@chainguard.dev>
Date: Tue, 15 Sep 2026 11:33:32 -0600
Subject: [PATCH 15/18] Revert "env: support --env0-from to read NUL separated
 vars from file"

This reverts commit adb4d9879962456d96bcde126cbba9356cc16f05.
NEWS left untouched.
---
 THANKS.in              |   1 -
 doc/coreutils.texi     |  13 -----
 src/env.c              |  78 ++++-------------------------
 tests/env/env0-from.sh | 111 -----------------------------------------
 tests/local.mk         |   1 -
 5 files changed, 10 insertions(+), 194 deletions(-)
 delete mode 100755 tests/env/env0-from.sh

diff --git a/THANKS.in b/THANKS.in
index 6a098897d..b1453be30 100644
--- a/THANKS.in
+++ b/THANKS.in
@@ -585,7 +585,6 @@ Soeren Sonnenburg                   sonnenburg@informatik.hu-berlin.de
 Solar Designer                      solar@owl.openwall.com
 Stanislav Ievlev                    inger@altlinux.ru
 Stavros Passas                      stabat@ics.forth.gr
-Stefan Klinger                      mail@stefan-klinger.de
 Stefan Vargyas                      stvar@yahoo.com
 Stéphane Chazelas                   Stephane_CHAZELAS@yahoo.fr
 Stephen Depooter                    sbdep@myrealbox.com
diff --git a/doc/coreutils.texi b/doc/coreutils.texi
index b00a265fd..60a4b19c1 100644
--- a/doc/coreutils.texi
+++ b/doc/coreutils.texi
@@ -17573,19 +17573,6 @@ environment.
 @optItemx{env,--ignore-environment,}
 Start with an empty environment, ignoring the inherited environment.
 
-@optItem{env,--env0-from,=@var{file}}
-Read NUL-terminated @samp{@var{name}=@var{value}} assignments from
-@var{file} and add them to the environment.  If @var{file} is
-@samp{-}, read from standard input.  The final assignment must be
-terminated by a NUL byte.
-
-Assignments from @var{file} replace inherited variables with the same
-name.  A later assignment in @var{file} takes precedence over an earlier
-one.  The @option{--ignore-environment} option starts with an empty
-environment before adding these assignments.  Variables specified by
-@option{--unset} are then removed, and command-line assignments take
-final precedence.
-
 @optItem{env,-C,@w{ }@var{dir}}
 @optItemx{env,--chdir,=@var{dir}}
 Change the working directory to @var{dir} before invoking @var{command}.
diff --git a/src/env.c b/src/env.c
index 8b9ba7741..9c7aef854 100644
--- a/src/env.c
+++ b/src/env.c
@@ -28,7 +28,6 @@
 #include "operand2sig.h"
 #include "printenv.h"
 #include "quote.h"
-#include "read-file.h"
 #include "sig2str.h"
 #include "xbinary-io.h"
 
@@ -45,9 +44,6 @@ static char const **usvars;
 static idx_t usvars_alloc;
 static idx_t usvars_used;
 
-/* Storage for variables passed to putenv from --env0-from.  */
-static char *env0_from_buffer;
-
 /* Annotate the output with extra info to aid the user.  */
 static bool dev_debug;
 
@@ -90,8 +86,7 @@ static char const shortopts[] = "+a:C:iS:u:v0" C_ISSPACE_CHARS;
    non-character as a pseudo short option, starting with CHAR_MAX + 1.  */
 enum
 {
-  ENV0_FROM_OPTION = CHAR_MAX + 1,
-  DEFAULT_SIGNAL_OPTION,
+  DEFAULT_SIGNAL_OPTION = CHAR_MAX + 1,
   IGNORE_SIGNAL_OPTION,
   BLOCK_SIGNAL_OPTION,
   LIST_SIGNAL_HANDLING_OPTION,
@@ -101,7 +96,6 @@ static struct option const longopts[] =
 {
   {"argv0", required_argument, NULL, 'a'},
   {"ignore-environment", no_argument, NULL, 'i'},
-  {"env0-from", required_argument, NULL, ENV0_FROM_OPTION},
   {"null", no_argument, NULL, '0'},
   {"unset", required_argument, NULL, 'u'},
   {"chdir", required_argument, NULL, 'C'},
@@ -139,10 +133,6 @@ Set each NAME to VALUE in the environment and run COMMAND.\n\
       oputs (_("\
   -i, --ignore-environment\n\
          start with an empty environment\n\
-"));
-      oputs (_("\
-      --env0-from=FILE\n\
-         set variables from NUL-delimited assignments in FILE\n\
 "));
       oputs (_("\
   -0, --null\n\
@@ -220,53 +210,6 @@ unset_envvars (void)
     }
 }
 
-/* Add ASSIGNMENT to the environment.  EQ points to its '=' byte.  */
-static void
-set_envvar (char *assignment, char *eq)
-{
-  devmsg ("setenv:   %s\n", assignment);
-
-  if (putenv (assignment))
-    {
-      *eq = '\0';
-      error (EXIT_CANCELED, errno, _("cannot set %s"), quote (assignment));
-    }
-}
-
-/* Merge the NUL-delimited assignments in FILE into the environment.  */
-static void
-set_envvars_from_file (char const *file)
-{
-  size_t size;
-  if (streq (file, "-"))
-    {
-      xset_binary_mode (STDIN_FILENO, O_BINARY);
-      env0_from_buffer = fread_file (stdin, RF_BINARY, &size);
-    }
-  else
-    env0_from_buffer = read_file (file, RF_BINARY, &size);
-
-  if (! env0_from_buffer)
-    error (EXIT_CANCELED, errno, _("cannot read %s"), quoteaf (file));
-  if (size && env0_from_buffer[size - 1] != '\0')
-    error (EXIT_CANCELED, 0, _("%s: file must end with a NUL byte"),
-           quotef (file));
-
-  char *end = env0_from_buffer + size;
-  for (char *assignment = env0_from_buffer; assignment < end; )
-    {
-      char *next = assignment + strlen (assignment) + 1;
-      char *eq = strchr (assignment, '=');
-      if (! eq)
-        error (EXIT_CANCELED, 0,
-               _("invalid variable specification %s in %s"),
-               quoteaf_n (0, assignment), quoteaf_n (1, file));
-
-      set_envvar (assignment, eq);
-      assignment = next;
-    }
-}
-
 /* Return a pointer to the end of a valid ${VARNAME} string, or NULL.
    'str' should point to the '$' character.
    First letter in VARNAME must be alpha or underscore,
@@ -828,7 +771,6 @@ main (int argc, char **argv)
   bool ignore_environment = false;
   bool opt_nul_terminate_output = false;
   char const *newdir = NULL;
-  char const *env0_from_file = NULL;
   char *argv0 = NULL;
 
   initialize_main (&argc, &argv);
@@ -862,9 +804,6 @@ main (int argc, char **argv)
         case '0':
           opt_nul_terminate_output = true;
           break;
-        case ENV0_FROM_OPTION:
-          env0_from_file = optarg;
-          break;
         case DEFAULT_SIGNAL_OPTION:
           parse_signal_action_params (optarg, true);
           parse_block_signal_params (optarg, false);
@@ -928,17 +867,20 @@ main (int argc, char **argv)
       static char *dummy_environ[] = { NULL };
       environ = dummy_environ;
     }
-
-  if (env0_from_file)
-    set_envvars_from_file (env0_from_file);
-
-  if (! ignore_environment || env0_from_file)
+  else
     unset_envvars ();
 
   char *eq;
   while (optind < argc && (eq = strchr (argv[optind], '=')))
     {
-      set_envvar (argv[optind], eq);
+      devmsg ("setenv:   %s\n", argv[optind]);
+
+      if (putenv (argv[optind]))
+        {
+          *eq = '\0';
+          error (EXIT_CANCELED, errno, _("cannot set %s"),
+                 quote (argv[optind]));
+        }
       optind++;
     }
 
diff --git a/tests/env/env0-from.sh b/tests/env/env0-from.sh
deleted file mode 100755
index b5cd1612b..000000000
--- a/tests/env/env0-from.sh
+++ /dev/null
@@ -1,111 +0,0 @@
-#!/bin/sh
-# Verify env --env0-from.
-
-# Copyright (C) 2026 Free Software Foundation, Inc.
-
-# This program is free software: you can redistribute it and/or modify
-# it under the terms of the GNU General Public License as published by
-# the Free Software Foundation, either version 3 of the License, or
-# (at your option) any later version.
-
-# This program is distributed in the hope that it will be useful,
-# but WITHOUT ANY WARRANTY; without even the implied warranty of
-# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-# GNU General Public License for more details.
-
-# You should have received a copy of the GNU General Public License
-# along with this program.  If not, see <https://www.gnu.org/licenses/>.
-
-. "${srcdir=.}/tests/init.sh"; path_prepend_ ./src
-print_ver_ env
-
-CU_ENV0_FROM_PARENT=parent
-CU_ENV0_FROM_REPLACE=parent
-CU_ENV0_FROM_REMOVE=parent
-export CU_ENV0_FROM_PARENT CU_ENV0_FROM_REPLACE CU_ENV0_FROM_REMOVE
-
-printf '%s\0' \
-  CU_ENV0_FROM_REPLACE=file \
-  CU_ENV0_FROM_NEW=new \
-  CU_ENV0_FROM_REMOVE=file >vars || framework_failure_
-
-# Without -i, merge the file with the inherited environment.  Apply -u
-# afterwards, and let command line assignments take final precedence.
-QUOTING_STYLE=literal env --env0-from=vars -u CU_ENV0_FROM_REMOVE \
-  CU_ENV0_FROM_REPLACE=operand >all || fail=1
-grep -E '^CU_ENV0_FROM_(NEW|PARENT|REMOVE|REPLACE)=' all |
-  LC_ALL=C sort >out || framework_failure_
-cat <<\EOF >exp || framework_failure_
-CU_ENV0_FROM_NEW=new
-CU_ENV0_FROM_PARENT=parent
-CU_ENV0_FROM_REPLACE=operand
-EOF
-compare exp out || fail=1
-
-# With -i (or a mere -), use only the assignments read from the file.
-QUOTING_STYLE=literal env --env0-from=vars -i >all || fail=1
-LC_ALL=C sort all >out || framework_failure_
-cat <<\EOF >exp || framework_failure_
-CU_ENV0_FROM_NEW=new
-CU_ENV0_FROM_REMOVE=file
-CU_ENV0_FROM_REPLACE=file
-EOF
-compare exp out || fail=1
-QUOTING_STYLE=literal env --env0-from=vars - >all || fail=1
-LC_ALL=C sort all >out || framework_failure_
-compare exp out || fail=1
-
-# Preserve values containing newlines, and accept an empty variable name.
-printf 'A=line 1\nline 2\0' >values || framework_failure_
-env -i --env0-from=values -0 >out || fail=1
-compare values out || fail=1
-printf '=value\0' >empty-name || framework_failure_
-env -i --env0-from=empty-name -0 >out || fail=1
-compare empty-name out || fail=1
-
-# A later record for the same name takes precedence.
-printf 'A=first\0B=second\0A=last\0' >duplicates || framework_failure_
-cat <<\EOF >exp || framework_failure_
-A=last
-B=second
-EOF
-QUOTING_STYLE=literal env -i --env0-from=duplicates >all || fail=1
-LC_ALL=C sort all >out || framework_failure_
-compare exp out || fail=1
-
-# Read '-' in binary mode from standard input.
-printf 'A=stdin\0' >exp || framework_failure_
-env -i --env0-from=- -0 <exp >out || fail=1
-compare exp out || fail=1
-
-# PATH loaded from the file must be used to find the command.
-mkdir bin || framework_failure_
-cat <<EOF >bin/env0-from-command || framework_failure_
-#!$SHELL
-echo found
-EOF
-chmod +x bin/env0-from-command || framework_failure_
-printf 'PATH=%s\0' "$PWD/bin" >path || framework_failure_
-echo found >exp || framework_failure_
-env -i --env0-from=path env0-from-command >out || fail=1
-compare exp out || fail=1
-
-# An empty file makes no changes, unless -i also requests an empty base.
-: >empty || framework_failure_
-env -i --env0-from=empty -0 >out || fail=1
-compare /dev/null out || fail=1
-
-# Every nonempty record must be a complete NUL-terminated assignment.
-printf 'A=unterminated' >invalid || framework_failure_
-returns_ 125 env --env0-from=invalid >out 2>err || fail=1
-printf 'not-an-assignment\0' >invalid || framework_failure_
-returns_ 125 env --env0-from=invalid >out 2>err || fail=1
-printf '\0' >invalid || framework_failure_
-returns_ 125 env --env0-from=invalid >out 2>err || fail=1
-returns_ 125 env --env0-from=does-not-exist >out 2>err || fail=1
-
-# Invalid -u names are still diagnosed after loading variables from a file.
-returns_ 125 env --env0-from=empty -u '' true || fail=1
-returns_ 125 env --env0-from=empty -u A=B true || fail=1
-
-Exit $fail
diff --git a/tests/local.mk b/tests/local.mk
index e21c1116d..8817d0e2c 100644
--- a/tests/local.mk
+++ b/tests/local.mk
@@ -357,7 +357,6 @@ all_tests =					\
   tests/date/date-tz.sh				\
   tests/misc/dircolors.pl			\
   tests/misc/dirname.pl				\
-  tests/env/env0-from.sh			\
   tests/env/env-null.sh				\
   tests/env/env-S.pl				\
   tests/env/env-S-script.sh			\
-- 
2.53.0

